learn/how-tos/Audithistory-using-crudListeners.md

```diff
@@ -0,0 +1,89 @@
+---
+title: "Record transactional History of an Entity"
+id: ""
+---
+---
+
+## Introduction
+
+In this document, we will learn how we can use [DB CRUD listeners](/learn/app-development/services/database-crud-event-listeners) to record the history(data change related details) of a specific entity/table in the logger table.
+Here, in our example, we use three EventListeners such as [EntityPostCreateEvent](https://github.com/wavemaker/wavemaker-app-runtime-services/blob/master/wavemaker-app-runtime-core/src/main/java/com/wavemaker/runtime/data/event/EntityPostCreateEvent.java), [EntityPostUpdateEvent](https://github.com/wavemaker/wavemaker-app-runtime-services/blob/master/wavemaker-app-runtime-core/src/main/java/com/wavemaker/runtime/data/event/EntityPostUpdateEvent.java) and [EntityPostDeleteEvent](https://github.com/wavemaker/wavemaker-app-runtime-services/blob/master/wavemaker-app-runtime-core/src/main/java/com/wavemaker/runtime/data/event/EntityPostDeleteEvent.java) with [@TransactionalEventListener](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/event/TransactionalEventListener.html) annotation.
+
+## Steps
+
+- Download the [EmployeeDB](/learn/assets/Audithistory-using-crudListeners/EmployeeDB_dump.sql) SQL file here and import this file in [DBTools](/learn/app-development/services/database-tools#db-scripts).
+
+- Now, connect to this database in your Application. In this use case, We use the EMPLOYEE_DETAILS_LOG table to insert transactional details of the EMPLOYEE table.
+
+- Create a java service in the project and add the below import statements.
+
+ ```java
+
+import org.springframework.beans.factory.annotation.Autowired;
+import com.wavemaker.runtime.security.SecurityService;
+import com.wavemaker.runtime.data.event.EntityPostCreateEvent;
+import com.wavemaker.runtime.data.event.EntityPostUpdateEvent;
+import com.wavemaker.runtime.data.event.EntityPostDeleteEvent;
+import org.springframework.context.event.EventListener;
+import com.<project_name>.employeedb.EmployeeDetailsLog;
+import com.<project_name>.employeedb.service.EmployeeDetailsLogService;
+import java.time.*;
+import com.<project_name>.employeedb.Employee;
+import com.wavemaker.runtime.data.exception.EntityNotFoundException;
+import org.springframework.transaction.event.TransactionalEventListener; 
+import org.springframework.transaction.event.TransactionPhase;
+```
+
+- Add below three methods into your java service. In this method, we get the logged-in user details from the security service and employee details from the event and insert them into the EMPLOYEE_DETAILS_LOG table.
+
+```java
+
+@TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
```

Comment:
we are performing a database operation in postEntityCreate after commit. It will again open a new transaction which will effect the performance. It is better to run this in normal event listener instead of transactional event listener.